### PR TITLE
*: set and use Flags to properly handle truncate error

### PR DESCRIPTION
**Pull request by coocood**
StatementContext is also updated according to Flags.

### CODE REVIEW

**Comment by coocood** [comment]
@zimulala @tiancaiamao @BusyJay PTAL

File: util/types/convert_test.go [hunk 1]
```diff
@@ -374,7 +373,7 @@ func testStrToUint(c *C, str string, expect uint64, truncateAsErr bool, expectEr
 
 func testStrToFloat(c *C, str string, expect float64, truncateAsErr bool, expectErr error) {
```
Comment (hanfei1991) [inline util/types/convert_test.go]: Why not pass IgnoreTruncate directly?

File: util/types/datum.go [hunk 2]
```diff
@@ -986,8 +986,12 @@ func (d *Datum) convertToMysqlDecimal(sc *variable.StatementContext, target *Fie
 		} else if frac != target.Decimal {
 			dec.Round(dec, target.Decimal)
 			if frac > target.Decimal {
-				if sc.TruncateAsError {
-					err = errors.Trace(ErrTruncated)
+				if !sc.IgnoreTruncate {
+					if sc.TruncateAsWarning {
+						sc.AppendWarning(ErrTruncated)
+					} else {
+						err = ErrTruncated
```
Comment (hanfei1991) [inline util/types/datum.go]: Why not trace error?

File: executor/statement_context_test.go [hunk 3]
```diff
@@ -20,19 +20,43 @@ import (
...
 	tk.MustExec("update sc set a = 3 where a > cast(1.1 as decimal)")
 	tk.MustQuery("select * from sc").Check(testkit.Rows("1", "3"))
+
+	tk.MustExec(strictModeSQL)
 	tk.MustExec("delete from sc")
-	tk.MustExec("set sql_mode = 'STRICT_TRANS_TABLES'")
 	tk.MustExec("insert sc values ('1.8'+1)")
 	tk.MustQuery("select * from sc").Check(testkit.Rows("3"))
+
+	// handle coprocessor flags, '1x' is an invalid data.
```
Comment (hanfei1991) [inline executor/statement_context_test.go]: handle -> Handle , same as next line

File: executor/statement_context_test.go [hunk 4]
```diff
@@ -20,19 +20,43 @@ import (
 	"github.com/pingcap/tidb/util/types"
 )
 
+const (
+	strictModeSQL    = "set sql_mode = 'STRICT_TRANS_TABLES'"
+	nonStrictModeSQL = "set sql_mode = ''"
+)
+
 func (s *testSuite) TestStatementContext(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
```
Comment (hanfei1991) [inline executor/statement_context_test.go]: Need we add defer here?

File: executor/statement_context_test.go (hunk 3 above)
Comment (hanfei1991) [inline executor/statement_context_test.go]: invalid data -> invalid int

File: distsql/xeval/eval.go [hunk 5]
```diff
@@ -118,3 +123,11 @@ func (e *Evaluator) evalIsNull(expr *tipb.Expr) (types.Datum, error) {
 	}
 	return types.NewIntDatum(0), nil
 }
+
+// FlagsToStatementContext creates a StatementContext from a `tipb.SelectRequest.Flags`.
+func FlagsToStatementContext(flags uint64) *variable.StatementContext {
+	sc := new(variable.StatementContext)
+	sc.IgnoreTruncate = flags&FlagIgnoreTruncate > 0
```
Comment (hanfei1991) [inline distsql/xeval/eval.go]: (flags & flagignoreTruncate) != 0 is more cleare

File: util/types/convert_test.go (hunk 1 above)
Comment (coocood) [inline util/types/convert_test.go]: Then too many test case need to change.

File: util/types/datum.go [hunk 6]
```diff
@@ -986,8 +986,12 @@ func (d *Datum) convertToMysqlDecimal(sc *variable.StatementContext, target *Fie
 		} else if frac != target.Decimal {
 			dec.Round(dec, target.Decimal)
 			if frac > target.Decimal {
-				if sc.TruncateAsError {
-					err = errors.Trace(ErrTruncated)
+				if !sc.IgnoreTruncate {
```
Comment (zimulala) [inline util/types/datum.go]: Extract a function to reduce the nested layers and it can be used in convert.go

File: distsql/xeval/eval.go [hunk 7]
```diff
@@ -35,6 +35,12 @@ const (
 	compareResultNull = -2
 )
 
+// Flags
+const (
+	FlagIgnoreTruncate    uint64 = 1
+	FlagTruncateAsWarning uint64 = 1 << 1
+)
```
Comment (disksing) [inline distsql/xeval/eval.go]:
```
const (
    FlagA = 1 << iota
    FlagB
)
```

Comment (coocood) [inline distsql/xeval/eval.go]: This flag is used by both TiDB and TiKV, we better make it explicit and consistent.

File: util/types/convert_test.go [hunk 8]
```diff
@@ -348,9 +347,9 @@ func (s *testTypeConvertSuite) TestConvertToString(c *C) {
 	}
 }
 
-func testStrToInt(c *C, str string, expect int64, truncateAsErr bool, expectErr error) {
+func testStrToInt(c *C, str string, expect int64, truncateAsError bool, expectErr error) {
```
Comment (hanfei1991) [inline util/types/convert_test.go]: Why this function's arg is AsError but the followings is AsErr ?

**Comment by coocood** [comment]
@hanfei1991 @disksing @ngaut PTAL

**Comment by hanfei1991** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**